### PR TITLE
Dynamic pricing (redux)

### PR DESCRIPTION
**Pull request by afiore**
Second implementation of [dynamic pricing](https://trello.com/c/hbnQF0nX/111-remove-all-hardcoding-of-prices-from-membership-front-end), this time using a configurable whitelist to map rate plans to Membership tiers.

Depends on the yet to be released [membership-commons 0.0.102-SNAPSHOT](https://github.com/guardian/membership-common/pull/113)

**Highlights** 

- Parsing and validation of Zuora catalog data has been centralised in `model.MembershipCatalog`. This is a simplified, "membership-specific" whereby pricing and rate plan data (including legacy ones) are now exposed to the rest of application. The mapping of rate plan ids to Membership tier plans happens in `touchpoint.ENV.conf`.
- Refined several controller method signatures based on the recently introduced differentiation between `PaidTier` and `TreeTier` (see PR 113 on membership commons).     
- Rewire controllers and views

### CODE REVIEW

**Comment by afiore** [comment]
@rtyley

File: frontend/conf/touchpoint.PROD.conf [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                }
+                partner {
+                    monthly="2c92a0fb4c5481dc014c69f95fce7240"
+                    yearly="2c92a0f94c54758b014c69f813bd39ec"
+                }
+                patron {
+                    monthly="2c92a0fb4c5481db014c69fb9118704b"
+                    yearly="2c92a0f94c547592014c69fb0c4274fc"
+                }
+                legacy {
```
Comment (rtyley) [inline frontend/conf/touchpoint.PROD.conf]: This hard-coding of legacy rateplanids makes me sad - I can understand the need for whitelisting the currently-on-sale rateplans - we don't wwant to sell the wrong thing - but I believe it should be possible to construct a model of 'legacy' tierplans from looking at all rateplans under products in the Product catalog that have a `Tier__c` field set. The billing period is also available from this information.


```
"id" : "2c92a0f9479fb46d0147d0155a3d5548",
    "sku" : "SKU-00000006",
    "name" : "Partner",
    "description" : "This is the description of the partner membership.",
    "ProductType__c" : "Membership",
    "ProductEnabled__c" : "True",
    "Entitlements__c" : "Events:PARTNER",
    "AcquisitionProfile__c" : "Paid",
    "Tier__c" : "Partner",
    "ProductLevel__c" : "200",
    "productRatePlans" : [ {
      "id" : "2c92a0f9479fb46d0147d0155cb15596",
      "status" : "Expired",
      "name" : "Partner - annual",
      "description" : "",
      "effectiveStartDate" : "2014-01-01",
      "effectiveEndDate" : "2015-06-23",
      "RatePlanType__c" : "Base",
      "DefaultTerm__c" : null,
      "TermType__c" : null,
      "PromotionCode__c" : null,
      "productRatePlanCharges" : [ {
        "id" : "2c92a0fb479fbd5d0147d01563184625",
        "name" : "Partner Membership - Annual",
        "type" : "Recurring",
        "model" : "FlatFee",
        "pricingSummary" : [ "GBP135" ],
        "billingDay" : "ChargeTriggerDay",
        "billingPeriod" : "Annual",
        "taxable" : true,
        "taxCode" : "Global Tax",
        "taxMode" : "TaxInclusive"
      } ]
    }, {
      "id" : "2c92a0f9479fb46d0147d0155ca15595",
      "status" : "Expired",
      "name" : "Partner - monthly",
      "description" : "",
      "effectiveStartDate" : "2014-01-01",
      "effectiveEndDate" : "2015-05-02",
      "RatePlanType__c" : "Base",
      "DefaultTerm__c" : null,
      "TermType__c" : null,
      "PromotionCode__c" : null,
      "productRatePlanCharges" : [ {
        "id" : "2c92a0fb479fbd5d0147d01562c9461a",
        "name" : "Partner Membership - Monthly",
        "type" : "Recurring",
        "model" : "FlatFee",
        "pricingSummary" : [ "GBP15" ],
        "billingDay" : "ChargeTriggerDay",
        "billingPeriod" : "Month",
        "taxable" : true,
        "taxCode" : "Global Tax",
        "taxMode" : "TaxInclusive"
      } ]
    },
```

File: frontend/app/model/MembershipCatalog.scala [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    def currenciesMatch(tierDetails: PaidTierDetails): Boolean =
+      tierDetails.monthlyPlanDetails.currencies == tierDetails.yearlyPlanDetails.currencies
+
+    object PublicPlans {
+      val ids = ratePlanIds
+      val friend = freeTierPlanDetails(FriendTierPlan.public, ids.friend)
+      val staff = freeTierPlanDetails(StaffPlan, ids.staff)
+      val supporterM = paidTierPlanDetails(PaidTierPlan.monthly(Supporter, Public), ids.supporterMonthly)
```
Comment (rtyley) [inline frontend/app/model/MembershipCatalog.scala]: I think you can reduce the number of declared variables here and just go with loops to reduce repetition.

File: frontend/app/model/MembershipCatalog.scala [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        .map(_.map { d => d.productRatePlanId -> d }.toMap)
+
+    def paidTierDetails(monthly: Val[PaidTierPlanDetails], yearly: Val[PaidTierPlanDetails]): Val[PaidTierDetails] =
+      (monthly |@| yearly)(PaidTierDetails).assert(currenciesMismatchErr)(currenciesMatch)
+
+    object Details {
+      private val P = PublicPlans
+      val friend = P.friend.map(FreeTierDetails)
+      val staff = P.staff.map(FreeTierDetails)
+      val supporter = paidTierDetails(P.supporterM, P.supporterY)
```
Comment (rtyley) [inline frontend/app/model/MembershipCatalog.scala]: You shouldn't have to say `paidTierDetails()` three times! Would really prefer fewer declared variables, better to loop over lists of tiers.

File: frontend/app/model/MembershipCatalog.scala [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        val msg = s"The catalog is inconsistent:\n" + errs.list.map(errorLine).mkString("\n")
+        throw MembershipCatalogException(msg)
+    }
+  }
+
+  def errorLine(p: (TierPlan, String)) = s"  - ${p._1}: ${p._2}"
+}
+
+trait MembershipCatalog {
+  def friend: FreeTierDetails
```
Comment (rtyley) [inline frontend/app/model/MembershipCatalog.scala]: Please try `lazy vals` instead of `def` - these are supposed to be immutable values!

Comment (ostapneko) [inline frontend/app/model/MembershipCatalog.scala]: Using `val` in trait does't seem to be a best practice, as it forces you to think hard about initialisation order. E.g. http://stackoverflow.com/questions/19642053/when-to-use-val-or-def-in-scala-traits

**Pull request merged**